File: PDR_EXT/src/main/cpp/PDR_EXT/src/Filter/MovingAverage.cpp
#include "Filter/MovingAverage.h"

MovingAverage::MovingAverage(int period)
        : period(period > 0 ? period : 0), sum(0.0f) {}

void MovingAverage::newData(float data) {
    sum += data;
    window.push(data);

    if (window.size() > period) {
        sum -= window.front();
        window.pop();
    }
}

float MovingAverage::getAvg() const {
    if (window.empty())
        return 0.0f;
    return sum / window.size();
}

void MovingAverage::reset() {
    while (!window.empty()) {
        window.pop();
    }
    sum = 0.0f;
}

File: maplocationlib/src/main/java/com/fifth/maplocationlib/sensors/MovingAverage.kt
package com.fifth.maplocationlib.sensors

import java.util.LinkedList
import java.util.Queue

class MovingAverage constructor(period : Int){
    private var window : Queue<Double> = LinkedList()
    private val period : Int = if (period > 0) period else 0
    private var sum : Double = 0.0

    fun newData(data: Float) {
        sum += data
        window.add(data.toDouble())
        if (window.size > period)
            sum -= window.poll()
    }

    fun getAvg() : Double {
        if (window.isEmpty())
            return 0.0
        else
            return sum / window.size
    }

    fun reset() {
        window.clear()
        sum = 0.0
    }
}